cpp/src/io/comp/debrotli.cu

```diff
@@ -307,51 +309,41 @@ static __device__ uint32_t getvlc(debrotli_state_s* s, const uint16_t* lut)
   return vlc;
 }
 
+static auto __device__ allocation_size(uint32_t bytes) { return (bytes + 7) & ~7; }
+
 /// Alloc bytes from the local (shared mem) heap
-static __device__ uint8_t* local_alloc(debrotli_state_s* s, uint32_t bytes)
+static __device__ void* local_alloc(debrotli_state_s* s, uint32_t bytes)
 {
-  int heap_used = s->heap_used;
-  int len       = (bytes + 7) >> 3;
-  if (heap_used + len <= s->heap_limit) {
-    uint8_t* ptr = reinterpret_cast<uint8_t*>(&s->heap[heap_used]);
-    s->heap_used = (uint16_t)(heap_used + len);
-    return ptr;
-  } else {
-    return nullptr;
-  }
+  auto const prev_heap_used = s->heap_used;
+  auto const len            = allocation_size(bytes);
+  if (prev_heap_used + len > s->heap_limit) { return nullptr; }
+  s->heap_used = prev_heap_used + len;
+  return &s->heap[prev_heap_used];
 }
 
 /// Shrink the size of the local heap, returns ptr to end (used for stack-like intermediate
 /// allocations at the end of the heap)
-static __device__ uint8_t* local_heap_shrink(debrotli_state_s* s, uint32_t bytes)
+static __device__ void* local_heap_shrink(debrotli_state_s* s, uint32_t bytes)
 {
-  int heap_used  = s->heap_used;
-  int heap_limit = s->heap_limit;
-  int len        = (bytes + 7) >> 3;
-  if (heap_limit - len >= heap_used) {
-    heap_limit -= len;
-    s->heap_limit = (uint16_t)heap_limit;
-    return reinterpret_cast<uint8_t*>(&s->heap[heap_limit]);
-  } else {
-    return nullptr;
-  }
+  auto const len = allocation_size(bytes);
+  if (s->heap_used + len > s->heap_limit) { return nullptr; }
+  s->heap_limit = s->heap_limit - len;
+  return &s->heap[s->heap_limit];
 }
 
 static __device__ void local_heap_grow(debrotli_state_s* s, uint32_t bytes)
 {
-  int len        = (bytes + 7) >> 3;
+  auto const len = allocation_size(bytes);
   int heap_limit = s->heap_limit + len;
   s->heap_limit  = (uint16_t)heap_limit;
 }
 
 /// Alloc memory from the fixed-size heap shared between all blocks (thread0-only)
-static __device__ uint8_t* ext_heap_alloc(uint32_t bytes,
-                                          uint8_t* ext_heap_base,
-                                          uint32_t ext_heap_size)
+static __device__ void* ext_heap_alloc(uint32_t bytes, void* ext_heap_base, uint32_t ext_heap_size)
 {
-  uint32_t len                = (bytes + 0xf) & ~0xf;
-  volatile uint32_t* heap_ptr = reinterpret_cast<volatile uint32_t*>(ext_heap_base);
-  uint32_t first_free_block   = ~0;
+  auto const len            = (bytes + 0xf) & ~0xf;
+  auto const heap_ptr       = static_cast<volatile uint32_t*>(ext_heap_base);
```

Comment:
This is `const` and `volatile` at the same time?

Reply:
The pointer is const, and the data it points to is volatile.
I can avoid auto here so the type becomes `uint32_t volatile* const`, which reads fine (right to left).

Reply:
Ok, that makes sense. I'd rather see the `const` on the left.